### ISSUE TITLE
Too many upsampling layers in decoder when reducing depth of FPN

### ISSUE THREAD

**Issue by JulienMaille** [qubvel/segmentation_models.pytorch#177]
https://github.com/qubvel/segmentation_models.pytorch/blob/af46195d9a805e621017e95aae6d02ca47df78b0/segmentation_models_pytorch/fpn/decoder.py#L99-L102

I tried a FPN-Resnet with 3 layers and I get a prediction 2 times too big when computing loss.
I believe you should adapt the number of `SegmentationBlock`, something like this
```python
self.seg_blocks = nn.ModuleList([
    SegmentationBlock(pyramid_channels, segmentation_channels, n_upsamples=n_upsamples)
    for n_upsamples in reversed(range(encoder_depth-1))
])
```

**Comment by qubvel** [comment]
Set another `upsampling` argument in model constructor

**Comment by JulienMaille** [comment]
Do you mean `upsampling` should be set to `encoder_depth-1`?

**Comment by qubvel** [comment]
You are right, we can reduce number of blocks, but another way is to reduce final upsampling.
Default is set to 4 for depth 5, for depth 4 - upsampling=2, for depth 3 - upsampling=1 to preserve input spatial shape.

**Comment by qubvel** [comment]
However, if you would like to increase speed it is better to remove upsampling blocks (like you suggested).

**Comment by JulienMaille** [comment]
> However, if you would like to increase speed it is better to remove upsampling blocks (like you suggested).

Yes I'm trying to work with lighter/faster models

**Comment by JulienMaille** [comment]
I don't want to open another issue, I have a similar question for Deeplab, with depth=3 I have an exception when trying to dilate removed stages. Is there another argument to pass, or should I add a check?
```python
        self.encoder.make_dilated(
            stage_list=[4, 5],
            dilation_list=[2, 4]
        )
```

**Comment by qubvel** [comment]
What kind of exception?

**Comment by JulienMaille** [comment]
Oh sorry, this is because of my patch to remove useless stages and avoid training/exporting them. With this patch it was trying to `make_dilated` a removed stage.
